File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/ByteBufferedInputStream.java
package titan.ast.runtime;

import java.io.IOException;
import java.io.InputStream;

/**
 * 类似BufferedInpustream功能的流.
 *
 * @author tian wei jun
 */
public class ByteBufferedInputStream {
  private static final int STANDARD_BUFFER_CAPACITY = 512;
  private static final int EOF = -1;

  public int nextReadIndex = 0;
  public int markNextReadIndex = 0;

  private int nextPos = 0;
  private int limit = 0;
  private int limitOfInvalidData = 0;
  private int mark = -1;

  private boolean isReadAllFromFile = false;

  private byte[] buffer;
  private final InputStream byteInputStream;

  /**
   * 带参构造.
   *
   * @param byteInputStream 所识别的文本输入流
   */
  public ByteBufferedInputStream(InputStream byteInputStream) {
    this.byteInputStream = byteInputStream;
    init();
  }

  /** 初始化. */
  public void init() {
    buffer = new byte[STANDARD_BUFFER_CAPACITY];
  }

  /**
   * 读取一个字符.
   *
   * @return 所读取的字符，若没有文本了，返回-1
   */
  public int read() throws IOException {
    if (nextPos < limit) { // 从缓冲中正常读取
      int read = buffer[nextPos++] & (int) 0xFF;
      ++nextReadIndex;
      return read;
    }
    // nextPos >= limit
    if (!isReadAllFromFile) { // 文件还没有读完就先填充，在尝试读取
      fillBuffer();
      return read();
    } else {
      return EOF;
    }
  }

  private void fillBuffer() throws IOException { // nextPos >= limit
    if (limit < buffer.length) {
      fillRemainder();
    } else {
      if (limitOfInvalidData > 0) {
        compact();
        fillRemainder();
      } else {
        fillByExpansion();
      }
    }
  }

  private void compact() {
    int moveCount = limit - limitOfInvalidData;
    System.arraycopy(buffer, limitOfInvalidData, buffer, 0, moveCount);
    nextPos = nextPos - limitOfInvalidData;
    mark = mark - limitOfInvalidData;
    limit = moveCount;
    limitOfInvalidData = 0;
  }

  private void fillRemainder() throws IOException {
    int countOfReaded = doRead(buffer, limit, buffer.length - limit);
    limit += countOfReaded;
  }

  private void fillByExpansion() throws IOException {
    int nsz = buffer.length + buffer.length;
    byte[] newBuffer = new byte[nsz];
    System.arraycopy(buffer, 0, newBuffer, 0, limit);

    // extend read
    int countOfNewRead = doRead(newBuffer, limit, nsz - limit);

    this.buffer = newBuffer;
    this.limit = this.limit + countOfNewRead;
  }

  /** mark为空 */
  public void reset() {
    if (mark < 0) {
      return;
    }
    nextPos = mark + 1;
    nextReadIndex = markNextReadIndex;
    if (nextPos >= limit) { // 数据全失效了
      nextPos = 0;
      limit = 0;
      limitOfInvalidData = 0;
    } else { // 还有可用数据
      limitOfInvalidData = nextPos;
    }
    mark = EOF;
  }

  public void mark() {
    this.mark = nextPos - 1;
    this.markNextReadIndex = nextReadIndex;
  }

  private int doRead(byte[] buffer, int offset, int len) throws IOException {
    int countOfRead = byteInputStream.read(buffer, offset, len);

    if (countOfRead == EOF) {
      isReadAllFromFile = true;
    }
    return Math.max(countOfRead, 0);
  }
}

File: runtime/cpp/titan-ast-runtime-lib/ByteBufferedInputStream.cpp
//
// Created by tian wei jun on 2022/12/2 0002.
//

#include "ByteBufferedInputStream.h"

const int ByteBufferedInputStream::STANDARD_BUFFER_CAPACITY = 512;

ByteBufferedInputStream::ByteBufferedInputStream()
    : nextReadIndex(0), eof(-1), nextPos(0), limit(0), markFlag(-1),
      markNextReadIndex(0), limitOfInvalidData(0), isReadAllFromFile(false),
      sizeOfBuffer(0), buffer(nullptr), byteInputStream(std::ifstream()) {}

ByteBufferedInputStream::~ByteBufferedInputStream() {
  delete[] buffer;
  buffer = nullptr;
  if (byteInputStream.is_open()) {
    byteInputStream.close();
  }
}

ReadResult ByteBufferedInputStream::read() {
  if (nextPos < limit) {
    int readedByte = buffer[nextPos++] & (int) 0xFF;
    ++nextReadIndex;
    return {true, readedByte};
  }
  if (!isReadAllFromFile) {// 文件还没有读完就先填充，在尝试读取
    ReadResult fillBufferResult = fillBuffer();
    if (!fillBufferResult.isOk) {
      return fillBufferResult;
    }
    return read();
  } else {
    return {true, eof};
  }
}

ReadResult ByteBufferedInputStream::fillBuffer() {
  if (limit < sizeOfBuffer) {
    return fillRemainder();
  } else {
    if (limitOfInvalidData > 0) {
      compact();
      return fillRemainder();
    } else {
      return fillByExpansion();
    }
  }
}

void ByteBufferedInputStream::compact() {
  int moveCount = limit - limitOfInvalidData;
  for (int i = 0; i < moveCount; i++) {
    buffer[i] = buffer[i + limitOfInvalidData];
  }
  nextPos = nextPos - limitOfInvalidData;
  markFlag = markFlag - limitOfInvalidData;
  limit = moveCount;
  limitOfInvalidData = 0;
}

ReadResult ByteBufferedInputStream::fillRemainder() {
  ReadResult doReadBufferResult =
      doReadBuffer(buffer, limit, sizeOfBuffer - limit);
  if (!doReadBufferResult.isOk) {
    return doReadBufferResult;
  }
  int countOfReaded = doReadBufferResult.data;
  limit += countOfReaded;
  return {true};
}

ReadResult ByteBufferedInputStream::fillByExpansion() {
  int nsz = sizeOfBuffer + sizeOfBuffer;
  byte *nBuffer = new byte[nsz];

  for (int i = 0; i < limit; i++) {
    nBuffer[i] = buffer[i];
  }
  // extend read
  ReadResult doReadBufferResult = doReadBuffer(nBuffer, limit, nsz - limit);
  if (!doReadBufferResult.isOk) {
    return doReadBufferResult;
  }
  int countOfNewRead = doReadBufferResult.data;

  delete[] this->buffer;
  this->buffer = nBuffer;
  this->sizeOfBuffer = nsz;
  this->limit = this->limit + countOfNewRead;
  return {true};
}

void ByteBufferedInputStream::reset() {
  if (markFlag < 0) {
    return;
  }
  nextPos = markFlag + 1;
  nextReadIndex = markNextReadIndex;
  if (nextPos >= limit) {// 数据全失效了
    nextPos = 0;
    limit = 0;
    limitOfInvalidData = 0;
  } else {// 还有可用数据
    limitOfInvalidData = nextPos;
  }
  markFlag = eof;
}

void ByteBufferedInputStream::mark() {
  this->markFlag = nextPos - 1;
  this->markNextReadIndex = nextReadIndex;
}

ReadResult ByteBufferedInputStream::doReadBuffer(byte *readBuffer, int offset,
                                                 int len) {
  char *base = reinterpret_cast<char *>(readBuffer + offset);
  int countOfRead = byteInputStream.read(base, len).gcount();
  if (byteInputStream.bad()) {
    return {false};
  }
  if (byteInputStream.eof()) {
    isReadAllFromFile = true;
  }
  if (countOfRead < 0) {
    countOfRead = 0;
  }
  return {true, countOfRead};
}

void ByteBufferedInputStream::clear() {
  nextReadIndex = 0;
  nextPos = 0;
  limit = 0;
  limitOfInvalidData = 0;
  markFlag = eof;

  isReadAllFromFile = false;

  if (byteInputStream.is_open()) {
    byteInputStream.close();
  }
  byteInputStream.clear();

  if (buffer) {
    delete[] buffer;
    buffer = nullptr;
    sizeOfBuffer = 0;
  }
}

ByteBufferedInputStreamInitResult
ByteBufferedInputStream::init(const std::string *sourceFilePath) {
  clear();

  if (!buffer) {
    buffer = new byte[STANDARD_BUFFER_CAPACITY];
    sizeOfBuffer = STANDARD_BUFFER_CAPACITY;
  }
  byteInputStream.open(*sourceFilePath, std::ios::in | std::ios::binary);
  if (!byteInputStream.is_open()) {
    return {false, "open source File error,path:'" + *sourceFilePath + "'"};
  }
  return {true, ""};
}
